core/src/kura.rs

```diff
@@ -4,258 +4,475 @@
+    /// read and write ops will fail. The most common reason this function
+    /// errors is because you forgot to call `create_files_if_they_do_not_exist`.
+    fn read_index_count(&self) -> Result<u64>;
+    /// Change the value returned by `read_index_count` aka the size of the
+    /// index file.
+    /// # Errors
+    /// Will fail if there is any IO error.
+    fn write_index_count(&mut self, new_count: u64) -> Result<()>;
+
+    /// Read from the data file in order to fill `dest_buffer` starting
+    /// from `start_location_in_data_file`.
+    /// # Errors
+    /// Will fail if there is any IO error.
+    fn read_block_data(
+        &self,
+        start_location_in_data_file: u64,
+        dest_buffer: &mut [u8],
+    ) -> Result<()>;
+    /// Write `block_data` into the data file starting at
+    /// `start_location_in_data_file`. Extend the file if
+    /// necessary.
+    /// # Errors
+    /// Will fail if there is any IO error.
+    fn write_block_data(
+        &mut self,
+        start_location_in_data_file: u64,
+        block_data: &[u8],
+    ) -> Result<()>;
+
+    /// Create the index and data files if they do not
+    /// already exist.
+    /// # Errors
+    /// Will fail if the any of the files don't exist
+    /// and couldn't be created.
+    fn create_files_if_they_do_not_exist(&mut self) -> Result<()>;
+
+    /// Above are the platform dependent functions.
+    /// Below are the platform independent functions that use the above functions.
+
+    /// Short hand for calling `read_block_indices` with a buffer of one.
+    /// # Errors
+    /// Simply passes on any errors from `read_block_indices`.
+    fn read_block_index(&self, block_height: u64) -> Result<(u64, u64)> {
+        let mut index = (0, 0);
+        self.read_block_indices(block_height, std::slice::from_mut(&mut index))?;
+        Ok(index)
+    }
+
+    /// Append `block_data` to this block store. First writing
+    /// the data to the data file and then creating a new index
+    /// for it.
+    /// # Errors
+    /// Fails if any of the required platform specific function operations
+    /// fail.
+    fn append_block_to_chain(&mut self, block_data: &[u8]) -> Result<()> {
+        let new_block_height = self.read_index_count()?;
+        let start_location_in_data_file = if new_block_height == 0 {
+            0
+        } else {
+            let (ultimate_block_start, ultimate_block_len) =
+                self.read_block_index(new_block_height - 1)?;
+            ultimate_block_start + ultimate_block_len
+        };
+
+        self.write_block_data(start_location_in_data_file, block_data)?;
+        self.write_block_index(
+            new_block_height,
+            start_location_in_data_file,
+            block_data.len() as u64,
+        )?;
+
+        Ok(())
+    }
+}
+
+/// An implementation of a block store for kura
+/// that uses rusts default file io. `std::fs`.
+pub struct StdFileBlockStore {
```

Comment:
I'd create an issue to find other potentially different `BlockStore` implementations.

Reply:
What do you mean by find? Seems more like something you would pass down from above. "We want to run on this system."

Reply:
I mean, in addition to `std::fs`, there could be other implementations. Some with fewer bells and whistles. We should create a meta-issue to gather those.